File: voidhack_agent_defender/rules.py
from __future__ import annotations

import re
import time
import urllib.parse
from dataclasses import dataclass, field

from .policy import Policy
from .schemas import CheckResult, Status, ToolCall

_URL_RE = re.compile(r"(?:[a-z][a-z0-9+.\-]*:)?//[^\s\"'<>]+", re.IGNORECASE)
_EMAIL_RE = re.compile(r"[A-Za-z0-9._%+\-]+@([A-Za-z0-9.\-]+\.[A-Za-z]{2,})")


@dataclass
class ToolFinding:
    tool_call_id: str
    tool_name: str
    status: Status
    reasons: list[str] = field(default_factory=list)
    hosts: list[str] = field(default_factory=list)
    secrets: list[str] = field(default_factory=list)
    arg_hits: list[str] = field(default_factory=list)


def _extract_hosts(arguments: str) -> list[str]:
    hosts: list[str] = []
    for url in _URL_RE.findall(arguments):
        netloc = urllib.parse.urlparse(url).netloc.lower()
        if netloc:
            hosts.append(netloc.split("@")[-1].split(":")[0])
    for match in _EMAIL_RE.finditer(arguments):
        hosts.append(match.group(1).lower())
    return hosts


def find_secrets(text: str, policy: Policy) -> list[str]:
    hits: list[str] = []
    for name, pattern in policy.compiled_secrets():
        if pattern.search(text or ""):
            hits.append(name)
    return hits


def check_arg_rules(tool: str, arguments: str, policy: Policy) -> list[str]:
    hits: list[str] = []
    for rule, pattern in policy.compiled_arg_rules():
        if rule.applies_to(tool) and pattern.search(arguments or ""):
            hits.append(rule.reason)
    return hits


def inspect_tool_call(tc: ToolCall, policy: Policy) -> ToolFinding:
    name = tc.function.name or "(unnamed)"
    args = tc.function.arguments or ""
    finding = ToolFinding(tool_call_id=tc.id, tool_name=name, status=Status.PASS)

    if not policy.tool_allowed(name):
        finding.status = Status.BLOCK
        verb = "denied" if name in policy.tool_denylist else "not on allowlist"
        finding.reasons.append(f"tool '{name}' is {verb}")

    hosts = _extract_hosts(args)
    finding.hosts = hosts
    bad_hosts = [host for host in hosts if not policy.host_allowed(host)]
    if bad_hosts:
        finding.status = Status.BLOCK
        finding.reasons.append(
            f"egress to non-allowlisted host(s): {', '.join(sorted(set(bad_hosts)))}"
        )

    secrets = find_secrets(args, policy)
    finding.secrets = secrets
    if secrets:
        finding.status = Status.BLOCK
        finding.reasons.append(f"secret(s) in tool args: {', '.join(secrets)}")

    arg_hits = check_arg_rules(name, args, policy)
    finding.arg_hits = arg_hits
    if arg_hits:
        finding.status = Status.BLOCK
        finding.reasons.extend(arg_hits)

    return finding


def check_tool_calls(
    tool_calls: list[ToolCall], policy: Policy
) -> tuple[list[ToolFinding], CheckResult]:
    started_at = time.perf_counter()
    findings = [inspect_tool_call(tool_call, policy) for tool_call in tool_calls]
    blocked = [finding for finding in findings if finding.status is Status.BLOCK]
    detail = (
        "; ".join(reason for finding in blocked for reason in finding.reasons)
        if blocked
        else f"{len(tool_calls)} tool call(s) within policy"
    )
    return findings, CheckResult(
        name="deterministic_rules",
        status=Status.BLOCK if blocked else Status.PASS,
        detail=detail,
        latency_ms=(time.perf_counter() - started_at) * 1000,
        meta={"blocked_tools": [finding.tool_name for finding in blocked]},
    )

File: agent-firewall-js/lib/rules.js
/**
 * Deterministic rules engine — sub-millisecond, zero network, never flaky.
 * Port of proxy/app/detect/rules.py to JavaScript.
 *
 * Checks: tool allow/deny, egress host allowlist, secret leaks in args,
 * and argument-level danger patterns (Trail of Bits class).
 */

// Matches any scheme URL or scheme-relative //host
const URL_RE = /(?:[a-z][a-z0-9+.\-]*:)?\/\/[^\s"'<>]+/gi;
const EMAIL_RE = /[A-Za-z0-9._%+\-]+@([A-Za-z0-9.\-]+\.[A-Za-z]{2,})/g;

/**
 * Extract hostnames from tool-call argument text.
 * @param {string} args
 * @returns {string[]}
 */
function extractHosts(args) {
  const hosts = [];
  // Reset lastIndex for global regexes
  URL_RE.lastIndex = 0;
  EMAIL_RE.lastIndex = 0;

  let m;
  while ((m = URL_RE.exec(args)) !== null) {
    try {
      const url = new URL(m[0]);
      const host = url.hostname.toLowerCase();
      if (host) hosts.push(host);
    } catch {
      // malformed URL — skip
    }
  }
  while ((m = EMAIL_RE.exec(args)) !== null) {
    hosts.push(m[1].toLowerCase());
  }
  return hosts;
}

/**
 * Check a list of tool calls against the policy.
 * @param {Array<{id: string, function: {name: string, arguments: string}}>} toolCalls
 * @param {import('./policy.js').Policy} policy
 * @returns {{findings: Array<Object>, blocked: boolean, detail: string}}
 */
export function checkToolCalls(toolCalls, policy) {
  const t0 = performance.now();
  const findings = toolCalls.map((tc) => inspectToolCall(tc, policy));
  const blocked = findings.filter((f) => f.status === "block");
  const detail = blocked.length
    ? blocked.map((f) => f.reasons.join("; ")).join(" | ")
    : `${toolCalls.length} tool call(s) within policy`;

  return {
    findings,
    blocked: blocked.length > 0,
    detail,
    latencyMs: performance.now() - t0,
  };
}

/**
 * Inspect a single tool call against policy rules.
 * @param {{id: string, function: {name: string, arguments: string}}} tc
 * @param {import('./policy.js').Policy} policy
 */
function inspectToolCall(tc, policy) {
  const name = tc.function?.name || "(unnamed)";
  const args = tc.function?.arguments || "";
  const finding = {
    toolCallId: tc.id,
    toolName: name,
    status: "allow",
    reasons: [],
    hosts: [],
    secrets: [],
    argHits: [],
  };

  // 1. Tool allow/deny
  if (!policy.toolAllowed(name)) {
    finding.status = "block";
    const verb = policy.tool_denylist.includes(name)
      ? "denied"
      : "not on allowlist";
    finding.reasons.push(`tool '${name}' is ${verb}`);
  }

  // 2. Egress allowlist
  const hosts = extractHosts(args);
  finding.hosts = hosts;
  const badHosts = hosts.filter((h) => !policy.hostAllowed(h));
  if (badHosts.length > 0) {
    finding.status = "block";
    const unique = [...new Set(badHosts)].sort();
    finding.reasons.push(
      `egress to non-allowlisted host(s): ${unique.join(", ")}`
    );
  }

  // 3. Secret leak in args
  const secrets = findSecrets(args, policy);
  finding.secrets = secrets;
  if (secrets.length > 0) {
    finding.status = "block";
    finding.reasons.push(`secret(s) in tool args: ${secrets.join(", ")}`);
  }

  // 4. Argument-level danger (Trail of Bits class)
  const argHits = checkArgRules(name, args, policy);
  finding.argHits = argHits;
  if (argHits.length > 0) {
    finding.status = "block";
    finding.reasons.push(...argHits);
  }

  return finding;
}

/**
 * Find secret patterns in text.
 * @param {string} text
 * @param {import('./policy.js').Policy} policy
 * @returns {string[]}
 */
function findSecrets(text, policy) {
  const hits = [];
  for (const { name, regex } of policy.secret_patterns) {
    if (regex.test(text || "")) {
      hits.push(name);
    }
    regex.lastIndex = 0; // reset for global-safe patterns
  }
  return hits;
}

/**
 * Check argument-level danger rules.
 * @param {string} toolName
 * @param {string} args
 * @param {import('./policy.js').Policy} policy
 * @returns {string[]}
 */
function checkArgRules(toolName, args, policy) {
  const hits = [];
  for (const rule of policy.arg_rules) {
    const applies =
      rule.tools.includes("*") || rule.tools.includes(toolName);
    if (applies && rule.regex.test(args || "")) {
      hits.push(rule.reason);
    }
    rule.regex.lastIndex = 0;
  }
  return hits;
}
